### PR TITLE
[TTAHUB-760] Always Maintain Region Filters in URL

### PR DESCRIPTION
**Pull request by AdamAdHocTeam**
## Description of change

We always want to maintain the region filters in the URL. This will ensure that when a URL filter is shared the user will be correctly notified if they are attempting to view a region they can't access.

## How to test

Test on both AR Landing and Regional Dashboard pages. 

Test for both a regular and Central Office user.

1. View the landing/dashboard page in a new browser. The URL should show all user regions.
2. Add a custom region filter. The URL should only show the region specified.
3. Remove the region filter. The URL should add back all regions the user can access.
4. Create a URL that has SOME of the regions the user can access. Paste it in a new browser window. You should see pills for the regions in the URL filter. 
5. Create a URL filter that has ALL of the regions the user can access. Paste it in a new browser window. There shouldn't be any filter pills for region but the URL should reflect all regions.

All preexisting filter functionality should still work as expected.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-684

### Every PR

- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [x] Meets accessibility standards (WCAG 2.1 Levels A, AA)
~- [ ] API Documentation updated~
~- [ ] Boundary diagram updated~
~- [ ] Logical Data Model updated~
~- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions~

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [x] Update JIRA ticket status

### CODE REVIEW

File: frontend/src/components/Modal.js [hunk 1]
```diff
@@ -37,18 +43,33 @@ const Modal = ({
       </div>
       <ModalFooter>
         <ButtonGroup>
-          <ModalToggleButton className={cancelButtonCss} data-focus="true" type="button" modalRef={modalRef} closer>
```
Comment (thewatermethod) [inline frontend/src/components/Modal.js:40]: Probably makes sense for this branch to track the permissions modal one, I see a bunch of duplicated code here (unless I'm confused, which is possible)

File: frontend/src/components/RegionPermissionModal.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+  const showFilterWithMyRegionAccess = () => {
+    showFilterWithMyRegions();
+    modalRef.current.toggleModal(false);
+  };
+
+  const smartSheetAccessLink = 'https://app.smartsheetgov.com/b/form/f0b4725683f04f349a939bd2e3f5425a';
+  const requestSmartSheetAccess = () => {
+    // Open link in new tab and close modal.
+    window.open(smartSheetAccessLink, '_blank');
```
Comment (thewatermethod) [inline frontend/src/components/RegionPermissionModal.js:44]: [Avoid resorting to window.open()](https://developer.mozilla.org/en-US/docs/Web/API/Window/open#avoid_resorting_to_window.open)

File: frontend/src/components/filter/FilterPanel.js [hunk 3]
```diff
@@ -1,22 +1,73 @@
...
+  onApplyFilters,
+  applyButtonAria,
+  allUserRegions,
+}) {
+  const [filtersToShow, setFiltersToShow] = useState([]);
+
+  useEffect(() => {
+    // Determine if filters contain all regions.
+    const passedRegionFilters = filters.filter((f) => f.topic === 'region').map((r) => {
+      if (isArray(r.query)) {
```
Comment (thewatermethod) [inline frontend/src/components/filter/FilterPanel.js:21]: As you were saying, I think if we actually want the URL to theoretically match from viewer to destination, we will have to include all visible regions in the URL

Comment (thewatermethod) [reply frontend/src/components/filter/FilterPanel.js:21]: ?region.nin[]=1&region.in[]=2

**Review by thewatermethod — COMMENTED** [review]
Seems to work when I tested it. A few tweaks as noted. If it were me, I'd probably check with the larger team before even getting into changing/fixing it

**Comment by AdamAdHocTeam** [comment]
> As a user with multiple regions I get a URL like `...?region.in[]=2&region.in[]=3&region.in[]=4&region.in[]=5&region.in[]=6...` on the landing page. This means we are getting all activity reports for "region 2 AND region 3..." and so on. This causes me to have zero activity reports loaded (since no activity report can be for multiple regions).

Right now my main concern going forward with this approach is how do we handle a region **NOT IN** filter. I think we would need to update the URL to contain all regions except the "NOT IN" region. Or we just leave the filter for this case as is for now?

Another option would be a copy filter button that automatically adds all other regions to the clipboard filter.

**Comment by jasalisbury** [comment]
> > As a user with multiple regions I get a URL like `...?region.in[]=2&region.in[]=3&region.in[]=4&region.in[]=5&region.in[]=6...` on the landing page. This means we are getting all activity reports for "region 2 AND region 3..." and so on. This causes me to have zero activity reports loaded (since no activity report can be for multiple regions).
> 
> Right now my main concern going forward with this approach is how do we handle a region **NOT IN** filter. I think we would need to update the URL to contain all regions except the "NOT IN" region. Or we just leave the filter for this case as is for now?
> 
> Another option would be a copy filter button that automatically adds all other regions to the clipboard filter.

Do we need to handle a user who doesn't have access to a "NOT IN" region filter? They don't have access to the region so they are seeing the correct data.

**Comment by AdamAdHocTeam** [comment]
> 

I think I misunderstood what you where saying here. 

Did you set a date on the alerts your expecting to see? If I create reports with multiple regions with no date the alerts don't show until I remove the date filter. There is a separate jira to fix this issue.

**Comment by AdamAdHocTeam** [comment]
> As a user with multiple regions I get a URL like `...?region.in[]=2&region.in[]=3&region.in[]=4&region.in[]=5&region.in[]=6...` on the landing page. This means we are getting all activity reports for "region 2 AND region 3..." and so on. This causes me to have zero activity reports loaded (since no activity report can be for multiple regions).

I think I misunderstood what you where saying here.

Did you set a date on the alerts your expecting to see? If I create reports with multiple regions with no date the alerts don't show until I remove the date filter. There is a separate jira to fix this issue.

**Comment by jasalisbury** [comment]
> > As a user with multiple regions I get a URL like `...?region.in[]=2&region.in[]=3&region.in[]=4&region.in[]=5&region.in[]=6...` on the landing page. This means we are getting all activity reports for "region 2 AND region 3..." and so on. This causes me to have zero activity reports loaded (since no activity report can be for multiple regions).
> 
> I think I misunderstood what you where saying here.
> 
> Did you set a date on the alerts your expecting to see? If I create reports with multiple regions with no date the alerts don't show until I remove the date filter. There is a separate jira to fix this issue.

You are correct, the date filter was causing no reports to show up, sorry about that!

File: frontend/src/components/filter/FilterPanel.js (hunk 3 above)
Comment (AdamAdHocTeam) [reply frontend/src/components/filter/FilterPanel.js:21]: I will discuss this with Josh.

File: frontend/src/components/RegionPermissionModal.js (hunk 2 above)
Comment (AdamAdHocTeam) [reply frontend/src/components/RegionPermissionModal.js]: This has been changed at the modal level to take a button. We now pass a link to open the smart sheet page.

File: frontend/src/components/filter/FilterPanel.js (hunk 3 above)
Comment (AdamAdHocTeam) [reply frontend/src/components/filter/FilterPanel.js:21]: Region not in has been removed.

**Review by jasalisbury — APPROVED** [review]
:rocket:

**Review by thewatermethod — APPROVED** [review]

**Pull request merged**